### PR TITLE
Cli 139

### PR DESCRIPTION
**Pull request by bright-poku**
### Description

This PR add `--offline-access` and `--token-expiration` to `apis create` and `apis update`

Screen Shot 2021-04-23 at 12 01 03 PM

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ X] The correct base branch is being used, if not `master`

### CODE REVIEW

File: internal/cli/apis.go [hunk 1]
```diff
@@ -38,6 +38,20 @@ var (
 		Help:       "Comma-separated list of scopes.",
 		IsRequired: true,
 	}
+	apiTokenLifetime = Flag{
+		Name:       "Token-Expiration",
```
Comment (Widcket) [inline internal/cli/apis.go:42]:
```suggestion
		Name:       "Token Expiration",
```

File: internal/cli/apis.go [hunk 2]
```diff
@@ -38,6 +38,20 @@ var (
...
+		LongForm:   "token-expiration",
+		ShortForm:  "t",
+		Help:       "The amount of time in seconds that the token will be valid after being issued",
+		IsRequired: true,
+	}
+	apiAllowOfflineAccess = Flag{
+		Name:       "Offline-Access",
```
Comment (Widcket) [inline internal/cli/apis.go:49]:
```suggestion
		Name:       "Offline Access",
```

File: internal/cli/apis.go [hunk 3]
```diff
@@ -38,6 +38,20 @@ var (
 		Help:       "Comma-separated list of scopes.",
 		IsRequired: true,
 	}
+	apiTokenLifetime = Flag{
+		Name:       "Token-Expiration",
+		LongForm:   "token-expiration",
+		ShortForm:  "t",
```
Comment (Widcket) [inline internal/cli/apis.go:44]:
```suggestion
		ShortForm:  "e",
```
For `expiration`

File: internal/cli/apis.go [hunk 4]
```diff
@@ -158,8 +174,9 @@ func createApiCmd(cli *cli) *cobra.Command {
 		Short: "Create a new API",
 		Long:  "Create a new API.",
 		Example: `auth0 apis create 
-auth0 apis create --name myapi 
-auth0 apis create -n myapi --identifier http://my-api`,
+auth0 apis create --name myapi
+auth0 apis create -n myapi --identifier http://my-api
+auth0 apis create -n myapi -i http://my-api --offline-access=true --token-expiration 6100 `,
```
Comment (Widcket) [inline internal/cli/apis.go:179]:
```suggestion
auth0 apis create -n myapi -i http://my-api --offline-access=true --token-expiration 6100`,
```

File: internal/cli/apis.go [hunk 5]
```diff
@@ -217,7 +246,9 @@ func updateApiCmd(cli *cli) *cobra.Command {
 		Long:  "Update an API.",
 		Example: `auth0 apis update 
 auth0 apis update <id|audience> 
-auth0 apis update <id|audience> --name myapi`,
+auth0 apis update <id|audience> --name myapi
+auth0 apis create -n myapi --token-expiration 6100
+auth0 apis create -n myapi -t 6100 --offline-access=true `,
```
Comment (Widcket) [inline internal/cli/apis.go:251]:
```suggestion
auth0 apis create -n myapi -t 6100 --offline-access=true`,
```

File: internal/cli/apis.go [hunk 6]
```diff
@@ -38,6 +38,20 @@ var (
...
+		LongForm:   "token-expiration",
+		ShortForm:  "t",
+		Help:       "The amount of time in seconds that the token will be valid after being issued",
+		IsRequired: true,
+	}
+	apiAllowOfflineAccess = Flag{
+		Name:       "Offline-Access",
+		LongForm:   "offline-access",
+		ShortForm:  "o",
+		Help:       "Allows issuance of refresh tokens for this entity",
```
Comment (Widcket) [inline internal/cli/apis.go:52]:
```suggestion
		Help:       "Allows issuance of refresh tokens for this API",
```

File: internal/display/apis.go [hunk 7]
```diff
@@ -46,16 +51,18 @@ type apiTableView struct {
...
 	Scopes     int
+	TokenLifetime int
+	AllowOfflineAccess bool
 
 	raw interface{}
 }
 
 func (v *apiTableView) AsTableHeader() []string {
-	return []string{"ID", "Name", "Identifier", "Scopes"}
+	return []string{"ID", "Name", "Identifier", "Scopes", "TokenLifetime", "AllowOfflineAccess"}
```
Comment (Widcket) [inline internal/display/apis.go:61]: Let's keep these new fields off the table view, as there's already enough info in there.

File: internal/display/apis.go [hunk 8]
```diff
@@ -16,16 +17,18 @@ type apiView struct {
...
 	Scopes     string
+	TokenLifetime int
+	AllowOfflineAccess bool
 
 	raw interface{}
 }
 
 func (v *apiView) AsTableHeader() []string {
-	return []string{"ID", "Name", "Identifier", "Scopes"}
+	return []string{"ID", "Name", "Identifier", "Scopes", "TokenLifetime", "AllowOfflineAccess"}
```
Comment (Widcket) [inline internal/display/apis.go:27]: Let's keep these new fields off the table view, as there's already enough info in there.

File: internal/display/apis.go (hunk 7 above)
Comment (Widcket) [inline internal/display/apis.go:61]: Let's keep these new fields off the table view, as there's already enough info in there.

File: internal/display/apis.go (hunk 8 above)
Comment (Widcket) [inline internal/display/apis.go:27]:
```suggestion
	return []string{"ID", "Name", "Identifier", "Scopes", "Token Lifetime", "Allow Offline Access"}
```

File: internal/cli/apis.go [hunk 9]
```diff
@@ -276,6 +316,8 @@ auth0 apis update <id|audience> --name myapi`,
 
 	apiName.RegisterStringU(cmd, &inputs.Name, "")
 	apiScopes.RegisterStringSliceU(cmd, &inputs.Scopes, nil)
+	apiAllowOfflineAccess.RegisterBool(cmd, &inputs.AllowOfflineAccess, true)
```
Comment (Widcket) [inline internal/cli/apis.go:319]:
```suggestion
	apiAllowOfflineAccess.RegisterBool(cmd, &inputs.AllowOfflineAccess, false)
```
This one shouldn't be required. I did not include this on the card, my bad.

File: internal/cli/apis.go [hunk 10]
```diff
@@ -38,6 +38,20 @@ var (
 		Help:       "Comma-separated list of scopes.",
 		IsRequired: true,
 	}
+	apiTokenLifetime = Flag{
+		Name:       "Token-Expiration",
+		LongForm:   "token-expiration",
+		ShortForm:  "t",
+		Help:       "The amount of time in seconds that the token will be valid after being issued",
+		IsRequired: true,
```
Comment (Widcket) [inline internal/cli/apis.go:46]:
```suggestion
		IsRequired: false,
```
Sorry, I did not include this on the card.

File: internal/cli/apis.go [hunk 11]
```diff
@@ -276,6 +316,8 @@ auth0 apis update <id|audience> --name myapi`,
 
 	apiName.RegisterStringU(cmd, &inputs.Name, "")
 	apiScopes.RegisterStringSliceU(cmd, &inputs.Scopes, nil)
+	apiAllowOfflineAccess.RegisterBool(cmd, &inputs.AllowOfflineAccess, true)
+	apiTokenLifetime.RegisterInt(cmd, &inputs.TokenLifetime, 0)
```
Comment (Widcket) [inline internal/cli/apis.go:320]:
```suggestion
	apiTokenLifetime.RegisterInt(cmd, &inputs.TokenLifetime, 86400)
```
`86400` (1 day) is the default value for new APIs

File: internal/cli/apis.go [hunk 12]
```diff
@@ -249,7 +280,16 @@ auth0 apis update <id|audience> --name myapi`,
 				return err
 			}
 
-			api := &management.ResourceServer{}
+			if err := apiTokenLifetime.Ask(cmd, &inputs.TokenLifetime, nil); err != nil {
```
Comment (Widcket) [inline internal/cli/apis.go:283]:
```suggestion
			if err := apiTokenLifetime.AskU(cmd, &inputs.TokenLifetime, nil); err != nil {
```
Use `AskU` for update command.

File: internal/cli/apis.go [hunk 13]
```diff
@@ -249,7 +280,16 @@ auth0 apis update <id|audience> --name myapi`,
...
+				return err
+			}
+
+			apiAllowOfflineAccess.AskBool(cmd, &inputs.AllowOfflineAccess, nil)
```
Comment (Widcket) [inline internal/cli/apis.go:287]:
```suggestion
			apiAllowOfflineAccess.AskBoolU(cmd, &inputs.AllowOfflineAccess, nil)
```
Use `AskBoolU` for update command

File: internal/display/apis.go (hunk 7 above)
Comment (bright-poku) [reply internal/display/apis.go:61]: sure I will make the changes

Comment (bright-poku) [reply internal/display/apis.go]: i think we should be good now

File: internal/cli/apis.go [hunk 14]
```diff
@@ -199,6 +224,8 @@ auth0 apis create -n myapi --identifier http://my-api`,
 	apiName.RegisterString(cmd, &inputs.Name, "")
 	apiIdentifier.RegisterString(cmd, &inputs.Identifier, "")
 	apiScopes.RegisterStringSlice(cmd, &inputs.Scopes, nil)
+	apiAllowOfflineAccess.RegisterBool(cmd, &inputs.AllowOfflineAccess, true)
+	apiTokenLifetime.RegisterInt(cmd, &inputs.TokenLifetime, 0)
```
Comment (Widcket) [inline internal/cli/apis.go:228]:
```suggestion
	apiTokenLifetime.RegisterInt(cmd, &inputs.TokenLifetime, 86400)
```
```suggestion
	apiTokenLifetime.RegisterInt(cmd, &inputs.TokenLifetime, 0)
```

File: internal/cli/apis.go [hunk 15]
```diff
@@ -199,6 +224,8 @@ auth0 apis create -n myapi --identifier http://my-api`,
 	apiName.RegisterString(cmd, &inputs.Name, "")
 	apiIdentifier.RegisterString(cmd, &inputs.Identifier, "")
 	apiScopes.RegisterStringSlice(cmd, &inputs.Scopes, nil)
+	apiAllowOfflineAccess.RegisterBool(cmd, &inputs.AllowOfflineAccess, true)
```
Comment (Widcket) [inline internal/cli/apis.go:227]:
```suggestion
	apiAllowOfflineAccess.RegisterBool(cmd, &inputs.AllowOfflineAccess, false)
```

Comment (bright-poku) [reply internal/cli/apis.go:227]: sorry i though those changes were commited. i will update them

File: internal/cli/apis.go [hunk 16]
```diff
@@ -249,7 +280,16 @@ auth0 apis update <id|audience> --name myapi`,
...
 			}
 
-			api := &management.ResourceServer{}
+			if err := apiTokenLifetime.AskU(cmd, &inputs.TokenLifetime, nil); err != nil {
+				return err
+			}
+
+			apiAllowOfflineAccess.AskBoolU(cmd, &inputs.AllowOfflineAccess, nil)
+
+			api := &management.ResourceServer{
```
Comment (Widcket) [inline internal/cli/apis.go:289]: This is an update command, please use the current existing values as the default unless new ones have been specified. Please take a look at the other params below for reference.

File: internal/cli/apis.go [hunk 17]
```diff
@@ -217,7 +247,9 @@ func updateApiCmd(cli *cli) *cobra.Command {
 		Long:  "Update an API.",
 		Example: `auth0 apis update 
 auth0 apis update <id|audience> 
-auth0 apis update <id|audience> --name myapi`,
+auth0 apis update <id|audience> --name myapi
+auth0 apis create -n myapi --token-expiration 6100
+auth0 apis create -n myapi -t 6100 --offline-access=true`,
```
Comment (Widcket) [inline internal/cli/apis.go:252]:
```suggestion
auth0 apis update -n myapi --token-expiration 6100
auth0 apis update -n myapi -t 6100 --offline-access=true`,
```

File: internal/cli/apis.go [hunk 18]
```diff
@@ -38,6 +39,20 @@ var (
 		Help:       "Comma-separated list of scopes.",
 		IsRequired: true,
 	}
+	apiTokenLifetime = Flag{
+		Name:       "Token Expiration",
+		LongForm:   "token-expiration",
+		ShortForm:  "e",
+		Help:       "The amount of time in seconds that the token will be valid after being issued",
+		IsRequired: false,
```
Comment (Widcket) [inline internal/cli/apis.go:47]:
```suggestion
```

File: internal/cli/apis.go [hunk 19]
```diff
@@ -158,8 +175,9 @@ func createApiCmd(cli *cli) *cobra.Command {
 		Short: "Create a new API",
 		Long:  "Create a new API.",
 		Example: `auth0 apis create 
-auth0 apis create --name myapi 
-auth0 apis create -n myapi --identifier http://my-api`,
+auth0 apis create --name myapi
+auth0 apis create -n myapi --identifier http://my-api
+auth0 apis create -n myapi -i http://my-api --offline-access=true --token-expiration 6100`,
```
Comment (Widcket) [inline internal/cli/apis.go:180]:
```suggestion
auth0 apis create --name myapi
auth0 apis create -n myapi --identifier http://my-api
auth0 apis create -n myapi --token-expiration 6100
auth0 apis create -n myapi -t 6100 --offline-access=true`,
```

**Comment by Widcket** [comment]
@bright-poku I've identified that some refactoring was needed for all the update cases to be handled correctly, and completed the work on #265. Thank you for this PR!

**Pull request closed without merging**